File: backend/templates/workflow-engine/actions/http-call.ts
/** HTTP call action dispatcher — makes HTTP requests via fetch(). */

import { ActionDispatcher } from './base';

export class HttpCallDispatcher extends ActionDispatcher {
  async execute(config: Record<string, unknown>): Promise<Record<string, unknown>> {
    const url = this.resolver.resolveString((config.url as string) ?? '');
    const method = ((config.method as string) ?? 'GET').toUpperCase();
    const headers = this.resolver.resolveDict(
      (config.headers as Record<string, unknown>) ?? {},
    ) as Record<string, string>;
    const body = config.body
      ? this.resolver.resolveDict(config.body as Record<string, unknown>)
      : undefined;

    console.log(`[workflow] HTTP ${method} ${url}`);

    const response = await fetch(url, {
      method,
      headers: { 'Content-Type': 'application/json', ...headers },
      body: body ? JSON.stringify(body) : undefined,
      signal: AbortSignal.timeout(30_000),
    });

    let responseBody: unknown;
    try {
      responseBody = await response.json();
    } catch {
      const text = await response.text();
      responseBody = text.slice(0, 2000);
    }

    return {
      action_type: 'http_call',
      status_code: response.status,
      response: responseBody,
      result: response.ok ? 'success' : 'error',
    };
  }
}

File: backend/runtime/actions/http_call.py
"""HTTP call action dispatcher — makes HTTP requests via httpx."""

import logging

from runtime.actions.base import ActionDispatcher

logger = logging.getLogger(__name__)


class HttpCallDispatcher(ActionDispatcher):
    """Makes an async HTTP request based on node config."""

    async def execute(self, config: dict) -> dict:
        import httpx

        url = self.resolver.resolve_string(config.get("url", ""))
        method = config.get("method", "GET").upper()
        headers = self.resolver.resolve_dict(config.get("headers", {}))
        body = self.resolver.resolve_dict(config.get("body", {})) if config.get("body") else None

        logger.info("HTTP %s %s", method, url)

        async with httpx.AsyncClient(timeout=30) as client:
            response = await client.request(
                method=method,
                url=url,
                headers=headers,
                json=body,
            )

        logger.info("HTTP response: %s %s", response.status_code, url)

        # Try to parse JSON response, fall back to text
        try:
            response_body = response.json()
        except Exception:
            response_body = response.text[:2000]

        return {
            "action_type": "http_call",
            "status_code": response.status_code,
            "response": response_body,
            "result": "success" if response.is_success else "error",
        }
